File: app/src/main/java/com/kareem/lifeos/BackgroundSemanticRuntime.java
package com.kareem.lifeos;

import android.content.Context;
import com.google.ai.edge.litertlm.Backend;
import com.google.ai.edge.litertlm.Conversation;
import com.google.ai.edge.litertlm.ConversationConfig;
import com.google.ai.edge.litertlm.Engine;
import com.google.ai.edge.litertlm.EngineConfig;
import com.google.ai.edge.litertlm.Message;
import java.io.File;

/** Process-local LiteRT-LM runtime shared by background semantics and grounded Ask LifeOS. */
final class BackgroundSemanticRuntime {
    private static Engine engine;private static String loadedPath="";
    private BackgroundSemanticRuntime(){}
    static synchronized String generate(Context context,String modelPath,String prompt)throws Exception{return run(context,modelPath,"You are LifeOS background semantic classification. Return only the requested JSON. Never invent facts, people, dates, urgency, actions, outcomes, or obligations.\n"+prompt);}
    static synchronized String generateFreeform(Context context,String modelPath,String prompt)throws Exception{return run(context,modelPath,prompt);}
    private static String run(Context context,String modelPath,String grounded)throws Exception{Context app=context.getApplicationContext();if(engine==null||!modelPath.equals(loadedPath)){reset();File cache=new File(app.getCacheDir(),"litertlm-background");cache.mkdirs();EngineConfig config=new EngineConfig(modelPath,new Backend.CPU(null,null),null,null,null,null,cache.getAbsolutePath());Engine fresh=new Engine(config);fresh.initialize();engine=fresh;loadedPath=modelPath;}Conversation conversation=null;try{conversation=engine.createConversation(new ConversationConfig());Message response=conversation.sendMessage(grounded);return response==null?"":response.toString().trim();}finally{if(conversation!=null)try{conversation.close();}catch(Throwable ignored){}}}
    static synchronized void reset(){if(engine!=null)try{engine.close();}catch(Throwable ignored){}engine=null;loadedPath="";}
}

File: app/src/main/java/com/kareem/lifeos/BackgroundSemanticRuntime.kt
package com.kareem.lifeos

import android.content.Context
import com.google.ai.edge.litertlm.Backend
import com.google.ai.edge.litertlm.Contents
import com.google.ai.edge.litertlm.ConversationConfig
import com.google.ai.edge.litertlm.Engine
import com.google.ai.edge.litertlm.EngineConfig
import com.google.ai.edge.litertlm.SamplerConfig
import java.io.File

/**
 * Process-local LiteRT-LM runtime for LifeOS background semantics.
 *
 * The engine is heavyweight, so it is initialized lazily on the dedicated BackgroundBrain
 * executor and reused across notification items. The no-think Qwen build is intentionally used
 * because LifeOS wants short structured JSON, not chain-of-thought text.
 */
object BackgroundSemanticRuntime {
    private var engine: Engine? = null
    private var loadedPath: String = ""

    @JvmStatic
    @Synchronized
    fun generate(context: Context, modelPath: String, prompt: String): String {
        val app = context.applicationContext
        val current = if (engine == null || loadedPath != modelPath) {
            try { engine?.close() } catch (_: Throwable) {}
            val cache = File(app.cacheDir, "litertlm-background").apply { mkdirs() }
            Engine(
                EngineConfig(
                    modelPath = modelPath,
                    backend = Backend.CPU(),
                    cacheDir = cache.absolutePath,
                )
            ).also {
                it.initialize()
                engine = it
                loadedPath = modelPath
            }
        } else engine!!

        val config = ConversationConfig(
            systemInstruction = Contents.of(
                "You are LifeOS background semantic classification. Return only the requested JSON. " +
                    "Never invent facts, people, dates, urgency, actions, outcomes, or obligations."
            ),
            samplerConfig = SamplerConfig(temperature = 0.0, topK = 1, topP = 1.0),
        )
        current.createConversation(config).use { conversation ->
            return conversation.sendMessage(prompt).contents.toString().trim()
        }
    }

    @JvmStatic
    @Synchronized
    fun reset() {
        try { engine?.close() } catch (_: Throwable) {}
        engine = null
        loadedPath = ""
    }
}
